### PR TITLE
refactor: change coveredCall to be more flexible, add swapExact helper

### PR DESCRIPTION
**Pull request by katelynsills**
This makes the `coveredCall` more flexible. Instead of trading a single kind of digital asset for another kind, any `give` and `want` can be traded, with multiple keywords under each. It also cuts down the coveredCall code from about 40 lines to 30 lines. The keywords no longer matter, so the buyer and seller can use different keywords with no issues. Furthermore, the creator and seller can use different keywords, as long as the issuers are known to Zoe at the time of the escrowing of the underlying assets.

### CODE REVIEW

File: packages/zoe/src/contractSupport/zoeHelpers.js
Comment (dtribble) [inline packages/zoe/src/contractSupport/zoeHelpers.js:229]: Based on this description, perhaps something that is more about that?  `swapAllocations`?

File: packages/zoe/src/contracts/coveredCall.js
Comment (dtribble) [inline packages/zoe/src/contracts/coveredCall.js]: I like putting this right at the top after the @ts-check. Does that seems like a plausible pattern?

File: packages/zoe/src/contracts/coveredCall.js
Comment (dtribble) [inline packages/zoe/src/contracts/coveredCall.js]: It would be nice to clarify the amount/payment issue here. The payout is not the strike price, since that's an amount. It's the "purchase payment" or something.

File: packages/zoe/src/contracts/coveredCall.js
Comment (dtribble) [inline packages/zoe/src/contracts/coveredCall.js:43]: Do we use backquotes for code references such as `offerResult`?

File: packages/zoe/src/contracts/coveredCall.js
Comment (dtribble) [inline packages/zoe/src/contracts/coveredCall.js]: is "assayable" still a word we use.  "verifiable"?

File: packages/zoe/src/contracts/coveredCall.js
Comment (dtribble) [inline packages/zoe/src/contracts/coveredCall.js:56]: Is it a bonus that they can use whatever keywords? We should figure out whether this is good or bad :).

File: packages/zoe/src/contracts/coveredCall.js
Comment (dtribble) [inline packages/zoe/src/contracts/coveredCall.js:70]: This is an uncomfortable pattern for debug info. Partly, try/catch that just adds info clutters the code. We pass one error message in but add this on the outside. should there be another error message argument?  an error object with properties for the various error modes?  This seems like we should discuss also with @erights

Comment (dtribble) [reply packages/zoe/src/contracts/coveredCall.js:70]: Expanding on this, these are tutorial messages. If `throw`n errors included codes, then it would be easier to add detailed documentation externally for what/why.

File: packages/zoe/src/contracts/coveredCall.js
Comment (dtribble) [inline packages/zoe/src/contracts/coveredCall.js]: should this be `timeAuthority` rather than `timerAuthority`?

File: packages/zoe/src/contracts/coveredCall.js
Comment (dtribble) [inline packages/zoe/src/contracts/coveredCall.js:13]: It wasn't obvious till later that the call can include multiple assets under different keywords. That's *really* cool and should be more obvious earlier!

File: packages/zoe/test/swingsetTests/zoe/vat-bob.js
Comment (dtribble) [inline packages/zoe/test/swingsetTests/zoe/vat-bob.js:100]: Here is where I figured out that the `coveredCall` could now support multiple assets :D

File: packages/zoe/test/unitTests/contracts/test-coveredCall.js
Comment (dtribble) [inline packages/zoe/test/unitTests/contracts/test-coveredCall.js:31]: For future reference, nothing about this scenario should require *Purses* as opposed to simply Payments.

File: packages/zoe/test/unitTests/contracts/test-coveredCall.js
Comment (dtribble) [inline packages/zoe/test/unitTests/contracts/test-coveredCall.js:136]: I'm a little worried that the additional flexibility at least about the StrikePrice will make coveredCall less recognizable as an option. Would we be better off having them be more narrowly structure to require a specific currency for the strikePrice.

File: packages/zoe/test/unitTests/contracts/test-coveredCall.js
Comment (dtribble) [inline packages/zoe/test/unitTests/contracts/test-coveredCall.js:170]: Is this testing that you can use arbitrary named?  or is it that you must use a different name for some underlying reason?

File: packages/zoe/test/unitTests/contracts/test-coveredCall.js
Comment (dtribble) [inline packages/zoe/test/unitTests/contracts/test-coveredCall.js:180]: API design issue: This whole pattern with `getPayout` with a named payout that could be wrong is scary. It seems very easy in many usages to drop a payment on the floor.  We need some mechanism or pattern to prevent that.

**Review by dtribble — COMMENTED** [review]
A few naming and design issues that should be resolved before commit.

File: packages/zoe/test/unitTests/contracts/test-coveredCall.js
Comment (katelynsills) [reply packages/zoe/test/unitTests/contracts/test-coveredCall.js:31]: Yup, this comes from our earlier test design when we were trying to imitate Alice as she would act in real life as much as possible.

File: packages/zoe/src/contractSupport/zoeHelpers.js
Comment (Chris-Hibbert) [reply packages/zoe/src/contractSupport/zoeHelpers.js:229]: or `swapAll`?

File: packages/zoe/test/unitTests/contracts/test-coveredCall.js
Comment (katelynsills) [reply packages/zoe/test/unitTests/contracts/test-coveredCall.js:136]: I think if we don't have this flexibility, we can't use the coveredCall as reusable "lego" as easily.

File: packages/zoe/test/unitTests/contracts/test-coveredCall.js
Comment (katelynsills) [reply packages/zoe/test/unitTests/contracts/test-coveredCall.js:170]: Arbitrarily named

File: packages/zoe/test/unitTests/contracts/test-coveredCall.js
Comment (katelynsills) [reply packages/zoe/test/unitTests/contracts/test-coveredCall.js:180]: Hmm, that's true. I'll turn this comment into an issue.

File: packages/zoe/src/contracts/coveredCall.js
Comment (Chris-Hibbert) [reply packages/zoe/src/contracts/coveredCall.js]: I thought eslint wanted `import '@something/...';` first

File: packages/zoe/src/contracts/coveredCall.js
Comment (Chris-Hibbert) [reply packages/zoe/src/contracts/coveredCall.js:43]: That's more useful for docs intended to be rendered by github. I think this one's intended to be displayed by the IDE. I don't think we have consistent expectations that all the IDEs support markdown in popup documentation.

File: packages/zoe/src/contracts/coveredCall.js
Comment (Chris-Hibbert) [reply packages/zoe/src/contracts/coveredCall.js]: I like assayable.

File: packages/zoe/src/contracts/coveredCall.js
Comment (Chris-Hibbert) [inline packages/zoe/src/contracts/coveredCall.js]: I don't think "Please check your payout" is the right terminology. Maybe "Please use" or "Please exercise"? "check" isn't active enough.

File: packages/zoe/test/unitTests/contracts/test-coveredCall.js
Comment (Chris-Hibbert) [reply packages/zoe/test/unitTests/contracts/test-coveredCall.js:136]: I think a specific keyword would be good. I'm not sure requiring the currency to be named when starting the contract is an improvement.

File: packages/zoe/src/zoeService/types.js
Comment (Chris-Hibbert) [inline packages/zoe/src/zoeService/types.js:189]: Is this so something else can have variant invitationDetails? I'd prefer the common case have the simpler name.

Comment (katelynsills) [reply packages/zoe/src/zoeService/types.js:189]: Nope, this is just that InvitationDetails can take custom details. The type was wrong before. StandardInvitationDetails will never be used alone

File: packages/zoe/test/unitTests/contracts/test-coveredCall.js
Comment (katelynsills) [reply packages/zoe/test/unitTests/contracts/test-coveredCall.js:136]: >  I'm not sure requiring the currency to be named when starting the contract is an improvement.

This is not true. Nothing needs to be named for anyone else. All keywords are independent of each other.

File: packages/zoe/src/contracts/coveredCall.js
Comment (katelynsills) [reply packages/zoe/src/contracts/coveredCall.js:56]: Yup, I think it is clearly a bonus, as this means that the coveredCall can be reused in other contexts, such as being called by another contract without having to remap the keywords.

File: packages/zoe/test/unitTests/contracts/test-coveredCall.js
Comment (katelynsills) [reply packages/zoe/test/unitTests/contracts/test-coveredCall.js:180]: Resolving this as it has been made into an issue

File: packages/zoe/src/contracts/coveredCall.js
Comment (erights) [inline packages/zoe/src/contracts/coveredCall.js:77]: Let's revisit this once https://github.com/Agoric/SES-shim/pull/447 lands and agoric-sdk uses it. Since we should revisit it then, I'm ok with leaving this bit of code as it is for now.

Given https://github.com/Agoric/SES-shim/pull/447 the closest corresponding code would be

```suggestion
      } catch (err) {
        assert.note(
          err, 
          details`Swap failed. Please make sure your offer has the same underlyingAssets and strikePrice as specified in the invitation details. The keywords should not matter.`,
        );
        throw err;
      }
```

This adds a hidden annotation to `err` so that, of err is ever logged to the improved console, this annotation is seen in the console log as well, and adjacent. If err never gets logged, for example because it is caught by a test case `throws`, then none of this appears in the console log.

Note the use of `details`. Arguments like `leftHasExitedMsg` can also be details rather than strings. This sends the actual values of the `${stuff}` substitution values to the console, whether or not you use `q` to reveal the stringification of these values to those with access to `err`.

File: packages/zoe/src/contracts/coveredCall.js
Comment (erights) [reply packages/zoe/src/contracts/coveredCall.js]: Not `exercise`. Especially not in a call options contract! Perhaps `Please collect the asset in your payout.` ?

File: packages/zoe/src/contractSupport/zoeHelpers.js
Comment (katelynsills) [reply packages/zoe/src/contractSupport/zoeHelpers.js:229]: I think `swapExact` is more precise, because this does not allow giving more than what is wanted by the other side.

File: packages/zoe/src/contractSupport/index.js
Comment (katelynsills) [reply packages/zoe/src/contractSupport/index.js:21]: Swap is different and still needed, I think. Swap only takes what is wanted by each side and let's the original offer keep the remainder. Swap doesn't need to be exact

File: packages/zoe/src/contracts/coveredCall.js
Comment (erights) [reply packages/zoe/src/contracts/coveredCall.js]: Yes, I like that it makes sense for both meanings of the word "authority.

File: packages/zoe/src/contracts/coveredCall.js
Comment (katelynsills) [reply packages/zoe/src/contracts/coveredCall.js]: If we use something from npm, eslint will require us to put those imports before a local one like ../../exported. So Chris is right that this pattern is not usable in general, although it is usable here because we don't import anything from npm.

File: packages/zoe/src/contracts/coveredCall.js
Comment (erights) [inline packages/zoe/src/contracts/coveredCall.js:86]: It is bizarre and surprising that the seller's choice of keywords is visible to the buyer here, without implication that the buyer should do anything with these keywords. But I think it actually helps the seller and buyer coordinate about the meaning of these assets. In any case, it doesn't hurt and is the most natural thing to do.

File: packages/zoe/src/zoeService/types.js
Comment (erights) [reply packages/zoe/src/zoeService/types.js:189]: Just now, reading the types, this helped me understand.  I would have come up with `Base*` rather than `Standard*`, but `Standard*` vs `Custom*` works better.

File: packages/zoe/src/contracts/coveredCall.js
Comment (katelynsills) [reply packages/zoe/src/contracts/coveredCall.js]: how about 'payments in the amount of the strike price'? purchase payment seems wrong, especially since the payout could include multiple payments

**Review by erights — APPROVED** [review]
I did not review any of the test files.

I did not review the extended comment text at the beginning of coveredCall.js

LGTM! I really like how well the coveredCall reads now!

File: packages/zoe/src/contracts/coveredCall.js
Comment (katelynsills) [reply packages/zoe/src/contracts/coveredCall.js]: I changed it to 'inspectable'

File: packages/zoe/src/contracts/coveredCall.js
Comment (katelynsills) [reply packages/zoe/src/contracts/coveredCall.js]: Done

File: packages/zoe/src/contracts/coveredCall.js
Comment (katelynsills) [reply packages/zoe/src/contracts/coveredCall.js:77]: I like this a lot

File: packages/zoe/src/contracts/coveredCall.js
Comment (katelynsills) [reply packages/zoe/src/contracts/coveredCall.js:86]: > It is bizarre and surprising that the seller's choice of keywords is visible to the buyer here, without implication that the buyer should do anything with these keywords. But I think it actually helps the seller and buyer coordinate about the meaning of these assets. In any case, it doesn't hurt and is the most natural thing to do.

Agreed. It is a bit jarring but unless we strip the keywords out and give an array of amounts, it must be so.

File: packages/zoe/src/contracts/coveredCall.js
Comment (katelynsills) [reply packages/zoe/src/contracts/coveredCall.js]: Done

File: packages/zoe/src/contracts/coveredCall.js
Comment (katelynsills) [reply packages/zoe/src/contracts/coveredCall.js:70]: Yup, I don't like the awkwardness of this either. But it seems like a long term thing that we should assess.

**Comment by katelynsills** [comment]
Ok, I believe I've tackled all of the near-term actionable comments. Since I got approval from @erights, I will merge after the tests pass.

**Review by erights — CHANGES_REQUESTED** [review]
Oops. I realized a problem after I had approved, but didn't submit this review before merge. But please fix anyway as it is a serious issue. It should be pure deletion.

File: packages/zoe/src/contracts/coveredCall.js
Comment (katelynsills) [reply packages/zoe/src/contracts/coveredCall.js:63]: These issuerKeywordRecords are still necessary. The offer which escrows the underlying assets needs to have the issuers saved in Zoe before the offer.

Comment (erights) [reply packages/zoe/src/contracts/coveredCall.js:63]: Kate and I just discussed. I misunderstood something. The change I suggest here would be incorrect. I withdraw the suggestion.

**Pull request merged**
